csrc/scheduler/vectorize_helper.cpp

```diff
@@ -64,63 +64,60 @@ ContiguousInnerDimensionsMapper::ContiguousInnerDimensionsMapper(
+  // can result in incorrect ordering
+  // NOTE: this is necessary to enable vectorization in
+  // NVFuserTest.FusionSegmentReduceSoftmax_CUDA
+  auto rfactor_domain = reference->getMaybeRFactorDomain();
+  auto filtered_ids = ids;
+  if (reference->isFusionInput()) {
+    auto remove_reduction = [](auto& container) {
+      container.erase(
+          std::remove_if(
+              container.begin(),
+              container.end(),
+              [](IterDomain* id) { return id->isReduction(); }),
+          container.end());
+    };
+    remove_reduction(rfactor_domain);
+    remove_reduction(filtered_ids);
+  }
 
   // Record while processing reference's information
   recording_ = true;
-  std::shared_ptr<Information> reference_information;
+  for (auto id : filtered_ids) {
+    addProjectedExtent(id, commonOrConstExtent(ca_map_, id));
+  }
 
   // Ordering of dimensions is important in this analysis, if an ordering is
   // contiguous in the reference, but not the target tensor views, then we
   // cannot consider that a contiguous merge dimension for vectorization.
-  std::vector<IterDomain*> reordered_rfactor;
-  for (auto id : reference->getMaybeRFactorDomain()) {
-    // Exclude reduction IDs if the reference is a fusion input as they
-    // don't manifest at all in the fusion. This simplifies the
-    // analysis in getContigMergeOfInnerSize, which only looks at
-    // non-reduction rfactor domains. Including reduction domains here
-    // can result in incorrect ordering
-    if (reference->isFusionInput() && id->isReduction()) {
-      continue;
-    }
-    // If not a fusion input, can this be a reduction domain?
-    TORCH_INTERNAL_ASSERT(
-        !id->isReduction(),
-        "Unexpected reduction domain: ",
-        id->toString(),
-        " of tensor ",
-        reference->toString());
-    if (std::find(reference_ids.begin(), reference_ids.end(), id) !=
-        reference_ids.end()) {
-      reordered_rfactor.push_back(id);
-      // Initiailze the extent for the mapped iter domain
-      addProjectedExtent(id, commonOrConstExtent(ca_map_, id));
-    } else if (!id->isBroadcast()) {
-      // Ignore broadcasts in the reference. Otherwise, remove non-contiguous
-      // IDs in the reference tensor as this is the contiguous mapper.
-      reordered_rfactor.clear();
+  auto projected_rfactor = projectId(filtered_ids, rfactor_domain);
+
+  // Need to somehow make sure projected_rfactor is indeed on the reference
+  // tensor, otherwise, getProjectedExtent seems to be complaining. I'm
+  // wondering if we should instead canonicalize the key for projected extent
+  auto map_rfactor = [this, &rfactor_domain](IterDomain* id) {
+    IterDomain* mapped_id = nullptr;
+    for (auto i : c10::irange(rfactor_domain.size())) {
+      if (this->ca_map_->idGraph().exactNodes().permissiveAreMapped(
+              rfactor_domain[i], id)) {
+        mapped_id = rfactor_domain[i];
+        break;
+      }
     }
-  }
```

Comment:
Looks like the new version has some difference regarding the initial reference IDs. In the current version, the reference IDs given to this function as a parameter are not directly used but they are reordered as the rfactor domain. I think that should actually be changed to the allocation domain, but use of the allocation domain in the scheduler has not yet started (@zasdfgbnm correct me if I'm wrong).

Also, in the current version the reference domains must be adjacent to each other except for broadcast domains.

In the new version, these properties don't seem to be enforced. Am I missing something?

Reply:
You are right that we no longer reorder them as the rFactor domain. I do think this is the correct behavior. When calling `ContiguousInnerDimensionsMapper`, you provide it with a vector of IterDomains `ids` in the reference tensor. The meaning for this vector is: "I will scheduler the reference tensor by merging all of these ids together, in the given order". And I don't really care if this order matches with the reference tensor's rFactor or root domain, because if the reference tensor is not a fusion input or output, why does it matter? What I indeed care is: is this order match with the fusion inputs/outputs? So when we start, we just project the given `ids` to the rFactor of the reference tensor, in the order of the given `ids`. Then we start propagating to the entire graph, setting the projected root and rFactor domain for each tensor. And when we want to compute the vectorization factor for an input/output tensor, we should first project from root or rFactor domain of that tensor to allocation domain of that tensor (which today we always assume allocation == rFactor), and then we check if the projected order matches with the allocation domain of that input/output tensor. If mismatch or if `contiguity == false` we just ignore left.

Reply:
> in the current version the reference domains must be adjacent to each other except for broadcast domains.
> In the new version, these properties don't seem to be enforced. Am I missing something?

That's true. But that restriction isn't necessary though.
When we project, we shouldn't care if the dimensions are "collapsible" on reference tensor. Because it doesn't matter how the reference tensor can vectorize. We blindly map the given ids onto each target tensor in the graph. Eventually when we compute aggregated contiguous inner domain sizes, we follow the rfactor domain there, since that dictates how the target tensor can be vectorized. https://github.com/NVIDIA/Fuser/blob/39b1a5a7e6de6caff89f886d21cd07aacf6558f4/csrc/scheduler/vectorize_helper.cpp#L756-L787

Reply:
regarding rfactor domain vs allocation domain.

I think we do want to move towards allocation domain, I'm leaning towards putting that up as a next step. Since we are not really looking at supporting arbitrary stride order at this moment.

Reply:
That makes sense. Thanks for the explanations!